### ISSUE TITLE
Phase A: add framework modules for ReplayBuffer, DQN, REINFORCE, CMA-ES, LSTM-ES

### ISSUE THREAD

**Issue by espenhk** [espenhk/gamer-ai#225]
Part of #224.

## Scope

Add five new modules under `framework/` that port the duplicated algorithm code currently sitting in `games/tmnf/policies.py`, `games/sc2/policies.py`, and `games/sc2/sc2_policies.py`. **No game files are touched in this phase.** Existing duplicates stay in place; the framework versions live alongside them until Phase C migrates each game.

This phase isolates the high-risk work — porting algorithm math — behind a clean interface boundary, with no risk of regressing live games.

### `framework/replay.py`

Port `ReplayBuffer` verbatim from `games/tmnf/policies.py:306` (identical to `games/sc2/policies.py:125`). Add `MaskedReplayBuffer` (6-tuple with `available_fn_ids`) from `games/sc2/policies.py:520`.

```python
class ReplayBuffer:
    """Circular buffer of (obs, action_idx, reward, next_obs, done) transitions."""
    def __init__(self, max_size: int) -> None: ...
    def add(self, obs, action_idx, reward, next_obs, done) -> None: ...
    def sample(self, batch_size: int, rng) -> tuple[np.ndarray, ...]: ...
    def __len__(self) -> int: ...

class MaskedReplayBuffer(ReplayBuffer):
    """Adds per-transition available-actions mask. Used by DQN with action masking."""
    def add(self, obs, action_idx, reward, next_obs, done, available_mask) -> None: ...
    def sample(self, batch_size: int, rng) -> tuple[np.ndarray, ...]: ...
```

### `framework/dqn.py`

Port `NeuralDQNPolicy` from `games/tmnf/policies.py:335`. Parameterise it on `obs_spec: ObsSpec` (instead of `BASE_OBS_DIM + n_lidar_rays`) and `discrete_actions: np.ndarray` (instead of the hardcoded `_DISCRETE_ACTIONS`). Add an optional `available_actions_fn: Callable[[dict], np.ndarray[bool]] | None` to absorb `SC2NeuralDQNPolicy` (`games/sc2/policies.py:568`) — when set, switch internal buffer to `MaskedReplayBuffer` and mask the Q-values before argmax.

```python
class DQNPolicy(BasePolicy):
    def __init__(
        self,
        obs_spec: ObsSpec,
        discrete_actions: np.ndarray,
        *,
        hidden_sizes: list[int] = (64, 64),
        replay_buffer_size: int = 10000,
        batch_size: int = 64,
        min_replay_size: int = 500,
        target_update_freq: int = 200,
        learning_rate: float = 0.001,
        epsilon_start: float = 1.0,
        epsilon_end: float = 0.05,
        epsilon_decay_steps: int = 5000,
        gamma: float = 0.99,
        available_actions_fn: Callable[[dict], np.ndarray] | None = None,
        seed: int | None = None,
    ) -> None: ...
```

Implements `save_trainer_state` / `load_trainer_state` once, writing the same `.npz` schema the per-game versions use today.

### `framework/reinforce.py`

Port `REINFORCEPolicy` from `games/tmnf/policies.py:953`. Parameterise on `obs_spec` and an `action_decoder: Callable[[int], np.ndarray]` (typically `lambda i: DISCRETE_ACTIONS[i]`).

Add a separate `TwoHeadREINFORCEPolicy` that mirrors `SC2REINFORCEPolicy` (`games/sc2/sc2_policies.py:567`) — fn_idx softmax + spatial softmax with independent sampling and combined loss. Both classes accept an optional `available_actions_fn` (for fn-head masking).

```python
class REINFORCEPolicy(BasePolicy):
    def __init__(self, obs_spec: ObsSpec, action_decoder: Callable[[int], np.ndarray],
                 *, output_dim: int, hidden_sizes=(64, 64), learning_rate=0.001,
                 gamma=0.99, entropy_coeff=0.01, baseline="running_mean",
                 available_actions_fn=None, seed=None) -> None: ...

class TwoHeadREINFORCEPolicy(BasePolicy):
    def __init__(self, obs_spec: ObsSpec, action_decoder: Callable[[int, int], np.ndarray],
                 *, fn_dim: int, spatial_dim: int, ...) -> None: ...
```

### `framework/cmaes.py`

Split into two classes:

```python
class CMAESDistribution:
    """Pure (μ/μ_w, λ)-CMA-ES math (Hansen 2016). No policy coupling.
    Operates only on θ ∈ R^n."""
    def __init__(self, n: int, lam: int, sigma: float, mean: np.ndarray | None = None,
                 seed: int | None = None) -> None: ...
    def sample(self) -> list[np.ndarray]: ...                # λ offspring
    def update(self, thetas: list[np.ndarray], fitnesses: np.ndarray) -> None: ...
    def save_state(self, path: str) -> None: ...
    def load_state(self, path: str) -> None: ...

class CMAESPolicy(BasePolicy):
    """Wraps CMAESDistribution + a game-injected parameter_decoder."""
    def __init__(self, obs_spec: ObsSpec,
                 parameter_decoder: Callable[[np.ndarray], BasePolicy],
                 flat_dim: int, *,
                 population_size=20, initial_sigma=0.3, eval_episodes=1,
                 seed=None) -> None: ...

    # Population-loop interface called by _greedy_loop_cmaes
    def sample_population(self) -> list[BasePolicy]: ...
    def update_distribution(self, rewards: np.ndarray) -> None: ...
    @property
    def champion_reward(self) -> float: ...
```

Verify byte-identical constants across the three existing copies (`games/tmnf/policies.py:712–722`, `games/sc2/policies.py:947–960`, `games/sc2/sc2_policies.py:1054–1062`) and port verbatim.

### `framework/lstm.py`

```python
class LSTMCore:
    """Single-layer LSTM cell + (h, c) state. Pure numpy. No action emission."""
    def __init__(self, obs_dim: int, hidden_size: int, seed=None) -> None: ...
    @property
    def flat_dim(self) -> int: ...
    def to_flat(self) -> np.ndarray: ...
    def from_flat(self, theta: np.ndarray) -> None: ...
    def forward(self, x: np.ndarray) -> np.ndarray: ...        # returns hidden state
    def reset(self) -> None: ...

class LSTMEvolutionPolicy(BasePolicy):
    """LSTMCore + game-injected head_decoder + isotropic-σ ES wrapper."""
    def __init__(self, obs_spec: ObsSpec,
                 head_decoder: Callable[[np.ndarray], np.ndarray],
                 head_param_count: int, *,
                 hidden_size=32, population_size=20, initial_sigma=0.05,
                 available_actions_fn=None, seed=None) -> None: ...
```

If `games/sc2/cnn_policy.py:SC2CNNEvolutionPolicy` (`:300`) uses the same isotropic-σ ES as the LSTM evolution wrappers, factor out a shared `IsotropicES` helper in `framework/lstm.py` (or a new `framework/es.py` — decide during implementation).

### Port existing TMNF-flavoured tests to the framework versions

Update the imports (no test-body changes needed in most cases):
- `tests/test_neural_dqn_policy.py` — import `DQNPolicy` from `framework.dqn`, parameterise with TMNF's `DISCRETE_ACTIONS` and obs spec.
- `tests/test_reinforce_policy.py` — import `REINFORCEPolicy` from `framework.reinforce`.
- `tests/test_cmaes_policy.py` — import `CMAESPolicy` from `framework.cmaes`.
- `tests/test_lstm_policy.py` — import `LSTMEvolutionPolicy` from `framework.lstm`.

The per-game test files (`tests/test_sc2_cmaes_policy.py` etc.) stay as-is in this phase — they still test the un-touched SC2 copies.

### New unit tests for `framework/cmaes.py:CMAESDistribution`

`tests/test_cmaes_distribution.py` — pure-math tests with no policy: seeded `sample()` returns expected shape, `update()` shifts mean toward the elite, σ adapts under the 1/5 success rule.

### Numeric equivalence tests (lock down the port)

`tests/test_framework_algorithm_equivalence.py` — for each ported algorithm, instantiate both the framework version and the corresponding game version with the same seed and feed identical (obs, reward) sequences. Assert byte-identical weights / θ / replay-buffer contents after N steps. This is the safety net for Phase C deletions.

### `tests/README.md`

Add sections for the new test files, per the CLAUDE.md test-suite README rule.

## Out of scope for Phase A

- Touching `games/*/policies.py` or `games/*/adapter.py`.
- Touching `framework/policies.py:BasePolicy` (the registry pattern lands in Phase B).
- Touching `framework/training.py` (the registry hookup lands in Phase B).
- Touching `framework/run_config.py` or `framework/game_adapter.py`.

## Verification

1. `python -m pytest tests/` — full suite green.
2. New `tests/test_framework_algorithm_equivalence.py` — framework and game versions produce byte-identical state on the same seed.
3. `python -c "from framework.replay import ReplayBuffer, MaskedReplayBuffer; from framework.dqn import DQNPolicy; from framework.reinforce import REINFORCEPolicy, TwoHeadREINFORCEPolicy; from framework.cmaes import CMAESDistribution, CMAESPolicy; from framework.lstm import LSTMCore, LSTMEvolutionPolicy; print('ok')"`.
4. `git diff --stat games/` — empty. No game files touched.

## Risks

- **Algorithm math drift between TMNF and SC2 copies.** Mitigation: the equivalence tests above. If any pair diverges, log the diff in the PR and choose the TMNF version as canonical (it's the older, more-exercised one).
- **LSTM ES wrapper shared with SC2 CNN policy.** Verify shape compatibility before extracting a shared `IsotropicES`; if `SC2CNNEvolutionPolicy` does anything materially different, leave it for a follow-up.

## Estimated size

- New code: ~800 lines across five framework modules.
- New tests: ~300 lines (equivalence + distribution unit tests).
- No deletions in this phase.
